monitors/new_end_of_run_monitor.py

```diff
@@ -0,0 +1,215 @@
+"""
+End of run monitor. Detects new runs that arrive and
+sends them off to the autoreduction service.
+"""
+
+import csv
+import logging
+import os
+import json
+from filelock import (FileLock, Timeout)
+
+from monitors.settings import (LAST_RUNS_CSV, CYCLE_FOLDER)
+
+from utils.clients.queue_client import QueueClient
+from utils.project.structure import get_log_file
+from utils.project.static_content import LOG_FORMAT
+
+# Setup logging
+EORM_LOG = logging.getLogger('end_of_run_monitor')
+EORM_LOG.setLevel(logging.INFO)
+
+FH = logging.FileHandler(get_log_file('end_of_run_monitor.log'))
+CH = logging.StreamHandler()
+FORMATTER = logging.Formatter(LOG_FORMAT)
+FH.setFormatter(FORMATTER)
+FH.setFormatter(FORMATTER)
+
+EORM_LOG.addHandler(FH)
+EORM_LOG.addHandler(CH)
+
+
+class InstrumentMonitorError(Exception):
+    """
+    Any fatal exception that occurs during execution of the
+    instrument monitor
+    """
+    pass
+
+
+class FileNotFoundError(Exception):
+    """
+    The run file couldn't be found. This may be because
+    of an inconsistency among DFS nodes.
+    """
+    pass
+
+
+def get_prefix_zeros(run_number_str):
+    """
+    Get the number of zeros prepended to a string
+    :param run_number_str: Run number as a string
+    :return: The zeros
+    """
+    zeros = ''
+    for num in run_number_str:
+        if num == '0':
+            zeros += '0'
+        else:
+            return zeros
+    return zeros
+
+
+class InstrumentMonitor(object):
+    """
+    Checks the ISIS archive for new runs on an instrument and submits them to ActiveMQ
+    """
+    def __init__(self, client, instrument_name):
+        self.client = client
+        self.instrument_name = instrument_name
+        self.summary_file = ""
+        self.last_run_file = ""
+        self.data_dir = ""
+        self.file_ext = ""
+
+    def read_instrument_last_run(self):
+        """
+        Read the last run recorded by the instrument from its lastrun.txt
+        :return: Last run on the instrument as a string
+        """
+        with open(self.last_run_file, 'r') as last_run_fp:
+            line_parts = last_run_fp.readline().split()
+            if len(line_parts) != 3:
+                raise InstrumentMonitorError("Unexpected last run file format for '{}'"
+                                             .format(self.last_run_file))
+        return line_parts
+
+    def read_rb_number_from_summary(self, run_number):
+        """
+        Loads the summary file and reads off the experiment RB number
+        :param run_number: Run number to lookup
+        :return: Experiment RB number
+        """
+        # Detect run number as a substring
+        with open(self.summary_file, 'rb') as summary_fp:
```

Comment:
What's the significance of `fp` suffix here?

Reply:
Stands for 'file pointer' from the C world. I wanted to distinguish it from the file name which is contained in summary_file.

Reply:
I'll change this because it's not very clear.